### PR TITLE
Optimize peak memory for flash _scaled_dot_product_attention_math (#139612)

### LINKED ISSUE #139612: Optimize peak memory for flash _scaled_dot_product_attention_math
### 🚀 The feature, motivation and pitch

  

Currently, `_scaled_dot_product_attention_math` uses `at::_safe_softmax(attn, -1)`, which increases memory usage compared to the previous `at::softmax(attn, -1)` implementation. We found that in some large model applications, the peak memory exceeds the memory capacity due to `_safe_softmax`, resulting in OOM.
The peak memory usage of `_scaled_dot_product_attention_math` is as follows:

**PT2.1**
image

pt=2.1.0a0+git7bcf7da: used=2360.94MB  

**PT2.6**  
image  
pt=2.6.0a0+gitf4ee5a2: used=3898.94MB

Test Scripts：  
```python
import torch
from torch.nn.attention import sdpa_kernel, SDPBackend
from torch.profiler import profile, ProfilerActivity

device = "cuda"

torch.backends.cuda.allow_fp16_bf16_reduction_math_sdp(True)
size = (1,32, 4096, 128)
q = torch.randn(size, device=device, dtype=torch.half, requires_grad=True)
k = torch.randn(size, device=device, dtype=torch.half, requires_grad=True)
v = torch.randn(size, device=device, dtype=torch.half, requires_grad=True)
with sdpa_kernel(SDPBackend.MATH):
    torch.cuda.memory._record_memory_history()
    res=torch.nn.functional.scaled_dot_product_attention(q, k, v)
    torch.cuda.memory._dump_snapshot("./sdpa_math_fwd_where_2.6.pickle")
    free, total = map(lambda x: x/2**20, torch.cuda.mem_get_info())
    used=total-free
    print(f'pt={torch.__version__}: {used=:0.2f}MB, {free=:0.2f}MB, {total=:0.2f}MB')
```

@drisspg @albanD 

### 

### Alternatives

I want to implement `_safe_softmax` in the following way, `at::where_out` reduces peak memory usage by reusing  memory
```cpp
Tensor _safe_softmax(
    const Tensor& self,
    int64_t dim,
    std::optional<ScalarType> dtype) {
  auto out = at::softmax(self, dim, dtype);
  const auto neg_inf = at::scalar_tensor(-std::numeric_limits<float>::infinity(), at::TensorOptions().dtype(out.dtype()).device(out.device()));
  const auto masked = self.eq(neg_inf);
  const auto masked_rows = all(masked, dim, true);
  const auto zero = at::scalar_tensor(0.0, at::TensorOptions().dtype(out.dtype()).device(out.device()));
  // reuse storage for out
 return at::where_out(out, masked_rows, zero, out);
}
```

The peak memory usage of `_scaled_dot_product_attention_math` is as follows:
image

pt=2.6.0a0+gitf4ee5a2: used=2874.94MB

### Additional context

_No response_

cc @ptrblck @msaroufim @drisspg @mikaylagawarecki

### PR DESCRIPTION
**Pull request by 1274085042**
Fixes #139612  

@drisspg @albanD

### CODE REVIEW

**Comment by drisspg** [comment]
This should be good lets check CI

**Comment by 1274085042** [comment]
@drisspg The CI checks seem to have completed

**Comment by janeyx99** [comment]
Is there a memory profile of before and after?

**Comment by drisspg** [comment]
Also For provenance, can you add the memory snapshots here as well from the issue?

**Comment by 1274085042** [comment]
@drisspg  @janeyx99   

Test Script:  
```python
import torch
from torch.nn.attention import sdpa_kernel, SDPBackend
from torch.profiler import profile, ProfilerActivity

device = "cuda"

torch.backends.cuda.allow_fp16_bf16_reduction_math_sdp(True)
size = (1,32, 4096, 128)
q = torch.randn(size, device=device, dtype=torch.half, requires_grad=True)
k = torch.randn(size, device=device, dtype=torch.half, requires_grad=True)
v = torch.randn(size, device=device, dtype=torch.half, requires_grad=True)
with sdpa_kernel(SDPBackend.MATH):
    torch.cuda.memory._record_memory_history()
    res=torch.nn.functional.scaled_dot_product_attention(q, k, v)
    torch.cuda.memory._dump_snapshot("./sdpa_math_fwd_where_2.6.pickle")
    free, total = map(lambda x: x/2**20, torch.cuda.mem_get_info())
    used=total-free
    print(f'pt={torch.__version__}: {used=:0.2f}MB, {free=:0.2f}MB, {total=:0.2f}MB')
```  

**before**  
image
pt=2.6.0a0+gitf4ee5a2: used=3898.94MB  

**after**  
image
pt=2.6.0a0+gitf4ee5a2: used=2874.94MB

**Review by drisspg — APPROVED** [review]
